tests/lms/views/decorators/test_h_api.py

```diff
@@ -0,0 +1,147 @@
+from pyramid.httpexceptions import HTTPBadRequest
+from pyramid.httpexceptions import HTTPGatewayTimeout
+import requests.exceptions
+
+from lms.views.decorators.h_api import create_h_user
+from lms.util import MissingToolConsumerIntanceGUIDError
+from lms.util import MissingUserIDError
+
+
+@pytest.mark.usefixtures("post", "util")
+class TestCreateHUser:
+    def test_it_400s_if_no_oauth_consumer_key_param(self, create_h_user, pyramid_request):
+        del pyramid_request.params["oauth_consumer_key"]
+
+        with pytest.raises(HTTPBadRequest, match="oauth_consumer_key"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_continues_to_the_wrapped_func_if_feature_not_enabled(self, create_h_user, pyramid_request, wrapped):
+        pyramid_request.params = {"oauth_consumer_key": "foo"}
+
+        returned = create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        wrapped.assert_called_once_with(pyramid_request, mock.sentinel.jwt)
+        assert returned == wrapped.return_value
+
+    def test_it_doesnt_use_the_h_api_if_feature_not_enabled(self, create_h_user, post, pyramid_request):
+        pyramid_request.params = {"oauth_consumer_key": "foo"}
+
+        create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        assert not post.called
+
+    def test_it_400s_if_generate_username_raises_MissingToolConsumerInstanceGUIDError(self, create_h_user, pyramid_request, util):
+        util.generate_username.side_effect = MissingToolConsumerIntanceGUIDError()
+
+        with pytest.raises(HTTPBadRequest, match="tool_consumer_instance_guid"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_username_raises_MissingUserIDError(self, create_h_user, pyramid_request, util):
+        util.generate_username.side_effect = MissingUserIDError()
+
+        with pytest.raises(HTTPBadRequest, match="user_id"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_provider_raises_MissingToolConsumerInstanceGUIDError(self, create_h_user, pyramid_request, util):
+        util.generate_provider.side_effect = MissingToolConsumerIntanceGUIDError()
+
+        with pytest.raises(HTTPBadRequest, match="tool_consumer_instance_guid"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_provider_unique_id_raises_MissingUserIDError(self, create_h_user, pyramid_request, util):
+        util.generate_provider_unique_id.side_effect = MissingUserIDError()
+
+        with pytest.raises(HTTPBadRequest, match="user_id"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_creates_the_user_in_h(self, create_h_user, post, pyramid_request):
+        create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        post.assert_called_once_with(
+            "https://example.com/api/users",
+            auth=("TEST_CLIENT_ID", "TEST_CLIENT_SECRET"),
```

Comment:
I feel daft, but I can't see where these values are coming from (`TEST_CLIENT_ID`, etc.)

Reply:
They're from the [default test settings](https://github.com/hypothesis/lms/blob/6e39b7e1e15be5df4048dcdf8b70fc1eff071461/tests/lms/conftest.py#L123-L172)